server/src/main/java/org/elasticsearch/search/aggregations/metrics/MedianAbsoluteDeviationAggregationBuilder.java

```diff
@@ -0,0 +1,140 @@
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregationBuilder.LeafOnly;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceParserHelper;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+public class MedianAbsoluteDeviationAggregationBuilder extends LeafOnly<ValuesSource.Numeric, MedianAbsoluteDeviationAggregationBuilder> {
+
+    public static final String NAME = "mad";
+
+    private static final ParseField COMPRESSION_FIELD = new ParseField("compression");
+
+    private static final ObjectParser<MedianAbsoluteDeviationAggregationBuilder, Void> PARSER;
+
+    static {
+        PARSER = new ObjectParser<>(NAME);
+        ValuesSourceParserHelper.declareNumericFields(PARSER, true, true, false);
+        PARSER.declareDouble(MedianAbsoluteDeviationAggregationBuilder::compression, COMPRESSION_FIELD);
+    }
+
+    public static MedianAbsoluteDeviationAggregationBuilder parse(String aggregationName, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new MedianAbsoluteDeviationAggregationBuilder(aggregationName), null);
+    }
+
+    private double compression = 1000d;
+
+    public MedianAbsoluteDeviationAggregationBuilder(String name) {
+        super(name, ValuesSourceType.NUMERIC, ValueType.NUMERIC);
+    }
+
+    public MedianAbsoluteDeviationAggregationBuilder(StreamInput in) throws IOException {
+        super(in, ValuesSourceType.NUMERIC, ValueType.NUMERIC);
+        compression = in.readDouble();
+    }
+
+    protected MedianAbsoluteDeviationAggregationBuilder(MedianAbsoluteDeviationAggregationBuilder clone,
+                                                        AggregatorFactories.Builder factoriesBuilder,
+                                                        Map<String, Object> metaData) {
+        super(clone, factoriesBuilder, metaData);
+        this.compression = clone.compression;
+    }
+
+    /**
+     * Returns the compression factor of the t-digest sketches used
+     */
+    public double compression() {
+        return compression;
+    }
+
+    /**
+     * Set the compression factor of the t-digest sketches used
+     */
+    public MedianAbsoluteDeviationAggregationBuilder compression(double compression) {
+        if (compression < 0.0) {
```

Comment:
Not really a MAD issue just something that caught my eye.  I know we allow zero as a valid compression in TDigest percentiles... but I'm wondering if we shouldn't allow it?

A quick skim of AVLTreeDigest shows `compression` used as a [denominator](https://github.com/tdunning/t-digest/blob/master/core/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L121) and in a [multiplication](https://github.com/tdunning/t-digest/blob/master/core/src/main/java/com/tdunning/math/stats/AVLTreeDigest.java#L154).  The agg still "works" with `0` compression, but I'm thinking that will cause a compression on each new point?  Should we allow that sort of behavior?

Reply:
+1 to not allowing compression of 0, it seems dangerous and trappy to me

Reply:
Definitely agree